### PR TITLE
chore(wallet): switch to chainstack proxies

### PR DESCRIPTION
**Pull request by onyb**
### Summary of changes

- `x-brave-key` is always extracted from buildflags, i.e., no more reading env vars
- switched to new RPC proxy endpoints following https://github.com/brave/devops/issues/11590
- ~remove Goerli network since it's been deprecated by EF~
- use chainstack RPC endpoint for BSC instead of public node
- use `x-brave-key` header for all Brave proxy endpoints

Resolves https://github.com/brave/brave-browser/issues/38872
Related to https://github.com/brave/devops/issues/11590

⚠️ Avalanche proxy is not correctly setup right now, tracking it here: https://github.com/brave/devops/issues/11914. For the time being we're using the public RPC until the proxy modification is live.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

The portfolio balances should load correctly without any console errors.

### CODE REVIEW

File: components/brave_wallet/browser/meld_integration_service.cc [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include "base/strings/stringprintf.h"
 #include "base/task/thread_pool.h"
 #include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
```
Comment (supermassive) [inline components/brave_wallet/browser/meld_integration_service.cc:23]: need that?

File: components/brave_wallet/browser/brave_wallet_constants.cc [hunk 2]
```diff
@@ -34,18 +34,29 @@ const std::string GetSardineNetworkName(const std::string& chain_id) {
...
+                  // Other chains
+                  {mojom::kBitcoinMainnet, "bitcoin-mainnet"}});
+
+  std::string chain_id_lower = base::ToLowerASCII(chain_id);
+  if (subdomains->contains(chain_id_lower)) {
+    return subdomains->at(chain_id_lower);
+  }
+
+  return "";
 }
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_constants.cc:61]: Let's make this an anonymous function in brave_wallet_utils.cc

File: browser/extensions/ethereum_remote_client_util.cc [hunk 3]
```diff
@@ -38,14 +38,7 @@ std::string GetInfuraProjectID() {
...
-  std::string brave_key(BUILDFLAG(BRAVE_SERVICES_KEY));
-  std::unique_ptr<base::Environment> env(base::Environment::Create());
-
-  if (env->HasVar("BRAVE_SERVICES_KEY")) {
-    env->GetVar("BRAVE_SERVICES_KEY", &brave_key);
-  }
-
-  return brave_key;
+  return BUILDFLAG(BRAVE_SERVICES_KEY);
 }
```
Comment (supermassive) [inline browser/extensions/ethereum_remote_client_util.cc:42]: There is some infura related stuff above, do we need it?

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 4]
```diff
@@ -929,29 +891,19 @@ mojom::NetworkInfoPtr GetChain(PrefService* prefs,
...
-    return GURL(endpoint);
-  }
-
-  auto subdomain = brave_wallet::GetInfuraSubdomainForKnownChainId(chain_id);
-  if (subdomain.empty()) {
-    return GURL();
-  }
-  return GURL(
-      base::StringPrintf("https://%s-infura.brave.com/", subdomain.c_str()));
+bool IsEndpointUsingBraveProxy(const GURL& url) {
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_utils.cc:894]: Let's name it IsEndpointUsingBraveWalletProxy

File: browser/extensions/ethereum_remote_client_util.cc (hunk 3 above)
Comment (onyb) [reply browser/extensions/ethereum_remote_client_util.cc:42]: We need for the old Brave Wallet which will still relies on an Infura project ID.

File: components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc [hunk 5]
```diff
@@ -169,8 +169,7 @@ std::string EndpointHost(const GURL& request_url) {
 
 bool ShouldThrottleEndpoint(const std::string& endpoint_host) {
   // Don't throttle requests if host matches brave proxy.
-  return EndpointHost(GURL(brave_wallet::kBitcoinMainnetRpcEndpoint)) !=
-         endpoint_host;
+  return !brave_wallet::IsEndpointUsingBraveWalletProxy(endpoint_host);
 }
```
Comment (supermassive) [inline components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc:173]: Let's switch EndpointHost to return URL:
```
GURL EndpointHost(const GURL& request_url) {
  DCHECK(request_url.is_valid());
  return request_url.GetWithEmptyPath();
}
```
and make `ShouldThrottleEndpoint` like
```
bool ShouldThrottleEndpoint(const GURL& endpoint_host) {
  // Don't throttle requests if host matches brave proxy.
  return !brave_wallet::IsEndpointUsingBraveWalletProxy(endpoint_host);
}
```

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 6]
```diff
@@ -929,29 +926,27 @@ mojom::NetworkInfoPtr GetChain(PrefService* prefs,
...
-GURL GetInfuraURLForKnownChainId(const std::string& chain_id) {
-  auto endpoint = brave_wallet::GetInfuraEndpointForKnownChainId(chain_id);
-  if (!endpoint.empty()) {
-    return GURL(endpoint);
-  }
+bool IsEndpointUsingBraveWalletProxy(const GURL& url) {
+  return url.DomainIs("wallet.brave.com") ||
+         url.DomainIs("wallet.bravesoftware.com") ||
+         url.DomainIs("wallet.s.brave.io");
+}
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_utils.cc:934]: Needs a tiny unit test

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 7]
```diff
@@ -1425,28 +1420,32 @@ std::vector<mojom::NetworkInfoPtr> GetAllKnownChains(PrefService* prefs,
 
   if (coin == mojom::CoinType::SOL) {
     for (const auto* network : GetKnownSolNetworks()) {
-      result.push_back(network->Clone());
+      result.push_back(
+          GetKnownChain(prefs, network->chain_id, mojom::CoinType::SOL));
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_utils.cc:1424]:
```
result.push_back(network->Clone());
```
Should be back here and below

File: components/brave_wallet/browser/meld_integration_service.cc (hunk 1 above)
Comment (onyb) [reply components/brave_wallet/browser/meld_integration_service.cc]: Actually we need this for `brave_wallet::MakeBraveServicesKeyHeaders()`

File: components/brave_wallet/browser/brave_wallet_utils.cc (hunk 7 above)
Comment (onyb) [reply components/brave_wallet/browser/brave_wallet_utils.cc]: https://github.com/brave/brave-core/pull/24045/commits/50cc22db2726e9b9430b03ea7a1e2ca96727c157

File: components/brave_wallet/browser/brave_wallet_utils.cc (hunk 4 above)
Comment (onyb) [reply components/brave_wallet/browser/brave_wallet_utils.cc]: https://github.com/brave/brave-core/pull/24045/commits/36124d274325debcadb2a917362ac02fde9a4c0d

File: components/brave_wallet/browser/brave_wallet_constants.cc (hunk 2 above)
Comment (onyb) [reply components/brave_wallet/browser/brave_wallet_constants.cc]: https://github.com/brave/brave-core/pull/24045/commits/da6f69ecd5b0910882b6429e9d0aec77190c977a

File: components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc (hunk 5 above)
Comment (onyb) [reply components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc]: https://github.com/brave/brave-core/pull/24045/commits/b1587bf3552918f346bf44ca2e9f78b35304dfa0

**Review by supermassive — APPROVED** [review]

**Review by josheleonard — APPROVED** [review]
Frontend ++

**Comment by srirambv** [comment]
Verification passed on

Brave | 1.69.18 Chromium: 126.0.6478.56 (Official Build) nightly (64-bit)
-- | --
Revision | 32b5c3ba39a2a4ecd9f07e3b44587971c23a6e40
OS | Linux

- Verified steps from PR
- Verified balances work as expected on Portfolio 
- Verified no console errors when loading Portfolio balance

**Pull request merged**